Crypto modern api

NelsonVides opened the pull request:
Solves #14, and also implements the SHA3 family of algorithms, which turned out to be a low hanging fruit after the API migration.

NelsonVides committed:
Add Clangd formatting rules

NelsonVides committed:
Add support for the SHA3 family

NelsonVides committed:
Reimplement using OpenSSL high-level API

This is the main rework. I initially wrote the standard functions for
SHA1, and only later refactored to extract all SHA1 specifics into macro
parameters, so that all other algorithms could be generated by the
macro.

It follows the same idea that the previous code, however it is not a
step-by-step refactoring but a rewrite from scratch.

NelsonVides committed:
Remove mix files, mix knows how to compile from rebar

DenysGonchar commented on README.md:
```diff
@@ -1,20 +1,22 @@
 
 `fast_pbkdf2` is an Erlang implementation of [PBKDF2][PBKDF2], where the algorithm is a carefully-optimised NIF that uses timeslicing and nif scheduling to respect the latency properties of the BEAM.
-All OTP versions from OTP18 have been tested manually and should work correctly, but on CI we support only from 21.3
+All OTP versions from OTP18 have been tested manually and should work correctly, but on CI we support only from 24.
```
is comment about OTP 18 still relevant?

NelsonVides replied:
Hmm... not really, not sure if anything have broken but what do I know, I'll remove that comment.

DenysGonchar commented on rebar.config:
```diff
@@ -32,7 +31,7 @@
   [
     {"(linux|solaris|freebsd|netbsd|openbsd|dragonfly|darwin|gnu)",
-         "CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -Wundef -Werror=undef -fPIC -I/opt/homebrew/include -I/usr/local/include"},
+         "CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -Werror -fPIC -I/opt/homebrew/include -I/usr/local/include"},
```
I believe the homebrew path is different for ARM and Intel based macbooks, but it's anyway a bad idea to add homebrew paths in rebar.config

chrzaszcz commented on c_src/fast_pbkdf2.c:
```diff
@@ -13,519 +14,432 @@
+    EVP_MD_CTX *outer;
+} HMAC_md_ctx;
+
+#define DECL_PBKDF2(_name, _blocksz, _hashsz, _iters_per_slot)                                     \
```
While I understand that this has changed completely, there was a useful description here. Without it, the code might be harder to understand.

chrzaszcz commented on c_src/fast_pbkdf2.c:
```diff
@@ -13,519 +14,512 @@
+    static inline int HMAC_INIT(_name)(HMAC_CTX_ROUND(_name) *restrict round_st,                   \
+                                       const EVP_MD *restrict type, const uint8_t *restrict key,   \
+                                       size_t nkey) {                                              \
+        uint8_t k[_blocksz];                                                                       \
```
Comment disappeared. Maybe keep it?

```
/* Prepare key: */
```

chrzaszcz commented on c_src/fast_pbkdf2.c:
```diff
@@ -13,519 +14,512 @@
+                                                                                                   \
+        return 0;                                                                                  \
+                                                                                                   \
+    err:                                                                                           \
```
In line 255, a similar label is named `error`. I'd keep them consistent, unless we want distinction, in which case I would propose `init_error` here.

chrzaszcz commented on c_src/fast_pbkdf2.c:
```diff
@@ -13,519 +14,512 @@
+                    goto error;                                                                    \
+                }                                                                                  \
+                                                                                                   \
+                /* XOR the outer hash into the result */                                           \
+                for (uint_fast8_t j = 0; j < _hashsz; ++j) {                                       \
+                    round_st->result[j] ^= round_st->Ublock[j];                                    \
+                }                                                                                  \
```
Was the previous code (which didn't use loops) faster here? Could this be the reason for some of the slowdown? Just wondering why you've changed that.

NelsonVides replied:
This is something that I'd expect the compiler to know how to optimise, pointers are declared as `restrict`, the compiler should know how to vectorise the loop hopefully 🤔  Will nevertheless have a look at the generated assembly, might be interesting 😄

NelsonVides replied:
Indeed: https://godbolt.org/z/95fE9nY9K
It's operating on big words (xmmword means 128bits at once) :)

NelsonVides replied:
And the code generated on my linux machine:
```asm
0x00000000000037dc <+268>: f3 0f 6f 40 20          movdqu 0x20(%rax),%xmm0 
0x00000000000037e1 <+273>: f3 0f 6f 48 60          movdqu 0x60(%rax),%xmm1 
0x00000000000037e6 <+278>: f3 0f 6f 50 70          movdqu 0x70(%rax),%xmm2 
0x00000000000037eb <+283>: f3 0f 6f 98 80 00 00 00 movdqu 0x80(%rax),%xmm3 
0x00000000000037f3 <+291>: 66 0f ef c1             pxor %xmm1,%xmm0 
0x00000000000037f7 <+295>: f3 0f 6f a0 90 00 00 00 movdqu 0x90(%rax),%xmm4 
0x00000000000037ff <+303>: 0f 11 40 20             movups %xmm0,0x20(%rax) 
0x0000000000003803 <+307>: f3 0f 6f 40 30          movdqu 0x30(%rax),%xmm0 
0x0000000000003808 <+312>: 66 0f ef c2             pxor %xmm2,%xmm0 
0x000000000000380c <+316>: 0f 11 40 30             movups %xmm0,0x30(%rax) 
0x0000000000003810 <+320>: f3 0f 6f 40 40          movdqu 0x40(%rax),%xmm0 
0x0000000000003815 <+325>: 66 0f ef c3             pxor %xmm3,%xmm0 
0x0000000000003819 <+329>: 0f 11 40 40             movups %xmm0,0x40(%rax) 
0x000000000000381d <+333>: f3 0f 6f 40 50          movdqu 0x50(%rax),%xmm0 
0x0000000000003822 <+338>: 66 0f ef c4             pxor %xmm4,%xmm0 
0x0000000000003826 <+342>: 0f 11 40 50             movups %xmm0,0x50(%rax)
```

NelsonVides committed:
Update README and badges

NelsonVides committed:
Update compilation flags in rebar.config

There was an issue in MacOS where in 15.x the dynamic linker would give
preference to the system's cached `libboringssl.dylib`, instead of the
OpenSSL's one put in the path. For the dynamic linker to resolve the
right library correctly, the static linker would need to add strict
resolution rules by being provided more strict flags.

NelsonVides committed:
Update ex_doc configuration in the rebar.config

NelsonVides committed:
Refactor CI entirely

NelsonVides commented:
There's some progress, now the failure is well handled and doesn't segfault nor aborts. But the failure's still there 🥲

NelsonVides committed:
Take into account error tuples returning from the NIF

chrzaszcz reviewed (commented):
I reviewed partially, because the code might change. Everything looks really good so far (I checked more than half of the C code).

DenysGonchar reviewed (approved):
looks good to me

chrzaszcz reviewed (approved):
Looks good 👌 I didn't analyse the algorithm itself that much, but the transformation from the previous code looks good to me.
